File: Sources/StatusMenuController.swift
import AppKit

/// Actions the menu bar can request. The controller depends on this narrow protocol
/// rather than the concrete app coordinator (Interface Segregation / Dependency Inversion).
protocol StatusMenuControllerDelegate: AnyObject {
    var isHeartGestureEnabled: Bool { get }
    func statusMenuDidRequestIncreaseOpacity()
    func statusMenuDidRequestDecreaseOpacity()
    func statusMenuDidToggleHeartGesture()
    func statusMenuDidRequestQuit()
}

/// Builds and owns the menu bar status item and its menu, forwarding clicks to its delegate.
final class StatusMenuController: NSObject {
    weak var delegate: StatusMenuControllerDelegate?

    private var statusItem: NSStatusItem?
    private var opacityItem: NSMenuItem?
    private var heartItem: NSMenuItem?

    func install() {
        let item = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        item.button?.image = GhostImageRenderer.menuBarImage(size: 18)
        item.menu = buildMenu()
        statusItem = item
    }

    func updateOpacity(_ opacity: Float) {
        opacityItem?.title = "Opacity: \(Int((opacity * 100).rounded()))%"
    }

    func updateHeartState(_ enabled: Bool) {
        heartItem?.state = enabled ? .on : .off
    }

    private func buildMenu() -> NSMenu {
        let menu = NSMenu()

        let opacity = NSMenuItem(title: "Opacity: —", action: nil, keyEquivalent: "")
        opacity.isEnabled = false
        menu.addItem(opacity)
        opacityItem = opacity

        menu.addItem(.separator())
        menu.addItem(makeItem("Increase Opacity", #selector(increase), key: "="))
        menu.addItem(makeItem("Decrease Opacity", #selector(decrease), key: "-"))

        let presets = NSMenuItem(title: "Set Opacity: ⌃⌥⌘ 1–9, 0 = 100%", action: nil, keyEquivalent: "")
        presets.isEnabled = false
        menu.addItem(presets)

        menu.addItem(.separator())
        let heart = makeItem("Heart Gesture", #selector(toggleHeart), key: "")
        heart.state = (delegate?.isHeartGestureEnabled ?? true) ? .on : .off
        menu.addItem(heart)
        heartItem = heart

        menu.addItem(.separator())
        menu.addItem(makeItem("Quit GlassFace", #selector(quit), key: "q"))
        return menu
    }

    private func makeItem(_ title: String, _ selector: Selector, key: String) -> NSMenuItem {
        let item = NSMenuItem(title: title, action: selector, keyEquivalent: key)
        if !key.isEmpty { item.keyEquivalentModifierMask = [.control, .option, .command] }
        item.target = self
        return item
    }

    @objc private func increase() { delegate?.statusMenuDidRequestIncreaseOpacity() }
    @objc private func decrease() { delegate?.statusMenuDidRequestDecreaseOpacity() }
    @objc private func toggleHeart() { delegate?.statusMenuDidToggleHeartGesture() }
    @objc private func quit() { delegate?.statusMenuDidRequestQuit() }
}

File: win/Sources/StatusMenuController.cs
using System;
using WinForms = System.Windows.Forms;

namespace GlassFace;

/// <summary>
/// Actions the tray menu can request. The controller depends on this narrow protocol rather
/// than the concrete app coordinator (Interface Segregation / Dependency Inversion).
/// </summary>
public interface IStatusMenuDelegate
{
    void StatusMenuDidRequestIncreaseOpacity();
    void StatusMenuDidRequestDecreaseOpacity();
    void StatusMenuDidRequestQuit();
}

/// <summary>
/// Builds and owns the system-tray icon and its menu, forwarding clicks to its delegate.
/// The Windows analog of the macOS menu-bar status item.
/// </summary>
public sealed class StatusMenuController : IDisposable
{
    public IStatusMenuDelegate? Delegate { get; set; }

    private WinForms.NotifyIcon? _icon;
    private WinForms.ToolStripMenuItem? _opacityItem;

    public void Install()
    {
        _icon = new WinForms.NotifyIcon
        {
            Icon = GhostImageRenderer.TrayIcon(),
            Text = "GlassFace",
            Visible = true,
            ContextMenuStrip = BuildMenu()
        };
    }

    public void UpdateOpacity(float opacity)
    {
        if (_opacityItem is not null)
            _opacityItem.Text = $"Opacity: {(int)Math.Round(opacity * 100)}%";
    }

    private WinForms.ContextMenuStrip BuildMenu()
    {
        var menu = new WinForms.ContextMenuStrip();

        _opacityItem = new WinForms.ToolStripMenuItem("Opacity: —") { Enabled = false };
        menu.Items.Add(_opacityItem);

        menu.Items.Add(new WinForms.ToolStripSeparator());
        menu.Items.Add(MakeItem("Increase Opacity\tCtrl+Alt+Shift +",
            (_, _) => Delegate?.StatusMenuDidRequestIncreaseOpacity()));
        menu.Items.Add(MakeItem("Decrease Opacity\tCtrl+Alt+Shift -",
            (_, _) => Delegate?.StatusMenuDidRequestDecreaseOpacity()));

        menu.Items.Add(new WinForms.ToolStripMenuItem("Set Opacity: Ctrl+Alt+Shift 1–9, 0 = 100%") { Enabled = false });

        menu.Items.Add(new WinForms.ToolStripSeparator());
        menu.Items.Add(MakeItem("Quit GlassFace\tCtrl+Alt+Shift Q",
            (_, _) => Delegate?.StatusMenuDidRequestQuit()));

        return menu;
    }

    private static WinForms.ToolStripMenuItem MakeItem(string title, EventHandler onClick)
    {
        var item = new WinForms.ToolStripMenuItem(title);
        item.Click += onClick;
        return item;
    }

    public void Dispose()
    {
        if (_icon is null) return;
        _icon.Visible = false;
        _icon.Dispose();
        _icon = null;
    }
}
